awx/main/utils/common.py

```diff
@@ -851,6 +851,25 @@ def get_mem_effective_capacity(mem_bytes):
 _workflow_manager = threading.local()
 
 
+@contextlib.contextmanager
+def task_manager_bulk_reschedule():
+    managers = [ScheduleTaskManager(), ScheduleWorkflowManager(), ScheduleDependencyManager()]
+    """Context manager to avoid submitting task multiple times."""
+    try:
+        for m in managers:
+            previous_flag = getattr(m.manager_threading_local, 'bulk_reschedule', False)
```

Comment:
When you reference `m.manager_threading_local`, that looks like it points to the global object `_task_manager`, because in the subclasses, like `ScheduleTaskManager`, that's what it passes in. So to rephrase this line, it would set `_task_manager.bulk_reschedule` to `True`, which would sacrifice the object-specific variable that I think you want.

You could unit test this and find out if I'm right about that or not.

`previous_flag` and `previous_value` have some scope confusion too. If you want, you could keep the `bulk_reschedule` on a global `threading.local()` object (that is, `_task_manager`), but importantly, `needs_scheduling` would need to be specific to the manager. There's considerable thought involved to get that right, and the core fix should be much more simply, which I'll point out in a subsequent suggestion.